File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.py
def solution(triangle):
    N = len(triangle)
    DP = [[0]*N for i in range(N)]
    DP[0][0] = triangle[0][0]
    
    for i in range(1, N):
        for j in range(i+1):
            # 맨 왼쪽 자리
            if j == 0:
                DP[i][j] = DP[i-1][j] + triangle[i][j]
            # 맨 오른쪽 자리
            elif j == i:
                DP[i][j] = DP[i-1][j-1] + triangle[i][j]
                
            else:
                DP[i][j] = max(DP[i-1][j], DP[i-1][j-1])+triangle[i][j]
                
    answer = max(DP[N-1])
    
    return answer
            

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
class Solution {
   public static int solution(int[][] triangle) {
	        int answer = 0;
	        int N = triangle.length; //5
	        int [][] dp = new int[N+1][N+1];
	        dp[0][0] = triangle[0][0];
	        
	        for(int i=1; i<N; i++) {
	        	for(int j=0; j<i+1; j++) {
	        		if(j==0) {
	        			dp[i][j] = dp[i-1][j]+triangle[i][j];
	        		}
	        		else if(j==i) {
	        			dp[i][j] = dp[i-1][j-1]+triangle[i][j]; //다음칸 1증가
	        		}
	        		else
	        			dp[i][j] = Math.max(dp[i-1][j], dp[i-1][j-1])+triangle[i][j];
	        	}
	        }
	        
	        
	        for(int i=0; i<N; i++) {
	        	answer = Math.max(answer, dp[N-1][i]);
	        }
	        return answer;
	    }
}